subtests/docker_cli/systemd/systemd.py

```diff
@@ -5,82 +5,124 @@
         self.sysd_action('daemon-reload')
-        self.sysd_action('start', self.config['sysd_unit_file'])
-        utils.wait_for(lambda: not is_port_free(port, host), 10)
-        time_from_socket = self.read_socket(host, port)
-        self.failif(not time_from_socket.isdigit(),
-                    "Data received from container is non-numeric: '%s'"
-                    % time_from_socket)
+        for act in 'start', 'status':
+            self.sysd_action(act, self.config['sysd_unit_file'])
 
     def cleanup(self):
-        super(systemd, self).cleanup()
-        filepath = self.config['sysd_unitf_dest']
+        super(systemd_base, self).cleanup()
+        filepath = self.sub_stuff['unitf_dst']
         if exists(filepath):
             self.sysd_action('stop', self.config['sysd_unit_file'])
             unlink(filepath)
             self.sysd_action('daemon-reload')
-        DockerImages(self).clean_all([self.config['name']])
+        DockerImages(self).clean_all([self.config['image_name']])
 
     @staticmethod
-    def sysd_action(action=None, sysd_unit=None):
+    def sysd_action(action, sysd_unit=None):
         command = 'systemctl {}'.format(action)
         if sysd_unit:
             command = '{} {}'.format(command, sysd_unit)
         utils.run(command, ignore_status=False)
 
-    @staticmethod
-    def read_socket(host, port):
-        max_read_tries = 10
-        num_read_try = max_read_tries
-        while num_read_try:
-            # using timeout 13-sec as it's 1-sec longer than two DNS timeouts
-            s = socket.create_connection((host, port), 13)
-            s.settimeout(2.0)
-            try:
-                time_from_socket = s.recv(512).strip()
-                if time_from_socket:
-                    s.shutdown(socket.SHUT_RDWR)
-                    return time_from_socket
-                num_read_try -= 1
-            except socket.timeout:
-                pass
-            time.sleep(1.0)
-        raise DockerTestError("Failed to read from socket with %d tries"
-                              % max_read_tries)
+    def write_unitfile(self, vars_dict):
+        tmpdirf = self.sub_stuff['tmpdirf']
+        tmp_write = '{}.tmp'.format(tmpdirf)
+        with open(tmpdirf, 'r') as srcfile:
+            lines = srcfile.readlines()
+            # if file is a unitfile, prepend line with 'Environment='
+            if '.service' in tmpdirf:
+                prep = 'Environment='
+            else:
+                prep = ''
+            # edit lines with matching string to dict keys
+            for key in vars_dict.keys():
+                ndl_found = next((ndl for ndl in lines if key in ndl), None)
+                if ndl_found is not None:
+                    indx = lines.index(ndl_found)
+                    lines[indx] = '{}{}{}\n'.format(prep, key, vars_dict[key])
+        with open(tmp_write, 'w') as outfile:
+            outfile.writelines(lines)
+        copyfile(tmp_write, tmpdirf)
```

Comment:
This still seems slow and clunky for what amounts to a search/replace of a single key/value.  I think I suggested before, there's no reason to iterate line-by-line. It's going to be much simpler (and faster) to read the entire file into a single string variable (it will preserve the newlines).  Then just do one search/replace across the entire thing all at once.  Then write the entire string out.

Also, I'm not sure why you've got the double-indirection of using an environment variable ``NAME``, and then referencing that from the other values.  If you're manipulating the file with code anyway, it's easier/faster to just search down all the tokens and replace them with duplicate values.  That's why I suggested just sticking the format tokens directly into your template unit files.

Unless I'm misunderstanding something, and you have to use the ``NAME`` environment for some reason.